src/python/pants/backend/kotlin/dependency_inference/kotlin_parser.py

```diff
@@ -0,0 +1,241 @@
+        )
+    elif len(source_files.files) == 0:
+        raise ValueError(
+            "analyze_scala_source_dependencies expects sources with exactly 1 source file, but found none."
+        )
+    source_prefix = "__source_to_analyze"
+    source_path = os.path.join(source_prefix, source_files.files[0])
+    processorcp_relpath = "__processorcp"
+    toolcp_relpath = "__toolcp"
+
+    (tool_classpath, prefixed_source_files_digest,) = await MultiGet(
+        Get(
+            ToolClasspath,
+            ToolClasspathRequest(artifact_requirements=_KOTLIN_PARSER_ARTIFACT_REQUIREMENTS),
+        ),
+        Get(Digest, AddPrefix(source_files.snapshot.digest, source_prefix)),
+    )
+
+    extra_immutable_input_digests = {
+        toolcp_relpath: tool_classpath.digest,
+        processorcp_relpath: processor_classfiles.digest,
+    }
+
+    analysis_output_path = "__source_analysis.json"
+
+    process_result = await Get(
+        FallibleProcessResult,
+        JvmProcess(
+            jdk=jdk,
+            classpath_entries=[
+                *tool_classpath.classpath_entries(toolcp_relpath),
+                processorcp_relpath,
+            ],
+            argv=[
+                "org.pantsbuild.backend.kotlin.dependency_inference.KotlinParserKt",
+                analysis_output_path,
+                source_path,
+            ],
+            input_digest=prefixed_source_files_digest,
+            extra_immutable_input_digests=extra_immutable_input_digests,
+            output_files=(analysis_output_path,),
+            extra_nailgun_keys=extra_immutable_input_digests,
+            description=f"Analyzing {source_files.files[0]}",
+            level=LogLevel.DEBUG,
+        ),
+    )
+
+    return FallibleKotlinSourceDependencyAnalysisResult(process_result=process_result)
+
+
+@rule(level=LogLevel.DEBUG)
+async def resolve_fallible_result_to_analysis(
+    fallible_result: FallibleKotlinSourceDependencyAnalysisResult,
+    process_cleanup: ProcessCleanupOption,
+) -> KotlinSourceDependencyAnalysis:
+    # TODO(#12725): Just convert directly to a ProcessResult like this:
+    # result = await Get(ProcessResult, FallibleProcessResult, fallible_result.process_result)
+    if fallible_result.process_result.exit_code == 0:
+        analysis_contents = await Get(
+            DigestContents, Digest, fallible_result.process_result.output_digest
+        )
+        analysis = json.loads(analysis_contents[0].content)
+        return KotlinSourceDependencyAnalysis.from_json_dict(analysis)
+    raise ProcessExecutionFailure(
+        fallible_result.process_result.exit_code,
+        fallible_result.process_result.stdout,
+        fallible_result.process_result.stderr,
+        "Kotlin source dependency analysis failed.",
+        process_cleanup=process_cleanup.val,
+    )
+
+
+@rule
+async def setup_kotlin_parser_classfiles(jdk: InternalJdk) -> KotlinParserCompiledClassfiles:
```

Comment:
Although we haven't extracted `javac` or `scalac` "tool compilation" `@rules` (#14737 started to do it, and would probably still be good to do), each new compiler is another opportunity to get things right the first time. So maybe worth extracting a kotlin compilation `@rule` now?

Reply:
Generally +1 to trying to re-factor things sooner.

Reply:
I'll do as a follow-up. I wanted to spend more of my limited spare time on getting the dependency inference working.